### ISSUE TITLE
New rule: Prefer short-circuiting assignment operators (TS 4+)

### ISSUE THREAD

**Issue by glen-84** [typescript-eslint/typescript-eslint#2263]
**Info:**

- https://github.com/tc39/proposal-logical-assignment/
- https://devblogs.microsoft.com/typescript/announcing-typescript-4-0-beta/

**Bad:**

```ts
opts.foo = opts.foo ?? 'bar'
```

**Good:**

```ts
opts.foo ??= 'bar'
```

(same for `||=` and `&&=`)

**Optional extras:**

Bad:

```ts
if (!a) {
  a = 'foo';
}
```

Good:

```ts
a ||= 'foo';
```

Bad:

```ts
opts.baz ?? (opts.baz = 'qux');
```

Good:

```ts
opts.baz ??= 'qux';
```

**Comment by bradzacher** [comment]
Will require #2260

**Comment by JoshuaKGoldberg** [comment]
Looks like this one is unblocked - I'll take it on! 🙌

**Comment by JoshuaKGoldberg** [comment]
Actually, on second thought, this really isn't a TypeScript-specific rule, as the operators are general JavaScript syntax. Should this be moved to ESLint core @bradzacher?

**Comment by bradzacher** [comment]
Probably could do yeah. They have support for the new logical operators!

**Comment by glen-84** [comment]
I'll move this issue (probably on Sunday or Monday).

**Comment by glen-84** [comment]
Closing in favour of https://github.com/eslint/eslint/issues/13689.
